# MAGs Processing Protocol

## 1. Bin Quality Check

### 1.1 CheckM

```bash
# Install
module load python/3.11.2
cd /project/def-careg421/ruizhang/software/
virtualenv --no-download checkm1.2.2-ENV
source /project/def-careg421/ruizhang/software/checkm1.2.2-ENV/bin/activate
pip install checkm-genome==1.2.2

# Install checkm db
mkdir /project/def-careg421/ruizhang/software/checkm1_db
cd /project/def-careg421/ruizhang/software/checkm1_db
wget https://zenodo.org/record/7401545/files/checkm_data_2015_01_16.tar.gz?download=1
mv checkm_data_2015_01_16.tar.gz?download=1 checkm_data_2015_01_16.tar.gz
tar -xvzf checkm_data_2015_01_16.tar.gz
export CHECKM_DATA_PATH=/project/def-careg421/ruizhang/software/checkm1_db

# Run CheckM
#!/bin/bash
#SBATCH -c 32
#SBATCH --mem=125G
#SBATCH -t 10:00:0
#SBATCH -J=gy2021_checkm1.sh

module load hmmer/3.3.2
module load prodigal/2.6.3
module load pplacer/1.1.alpha19
source /project/def-careg421/ruizhang/software/checkm1.2.2-ENV/bin/activate
mkdir /project/def-careg421/ruizhang/guiyang2021/checkm1/
export CHECKM_DATA_PATH=/project/def-careg421/ruizhang/software/checkm1_db
list="HX_surface HX_rhizosphere GX_surface GX_rhizosphere SK_rhizosphere"
for site in $list
do
input=/project/def-careg421/ruizhang/2021_guiyang/raw_bins/dastool/${site}/bins
output=/project/def-careg421/ruizhang/2021_guiyang/raw_bins/checkm1/${site}
checkm lineage_wf -t 32 -x fasta $input $output
done

# Combine CheckM results
cat /project/def-careg421/ruizhang/2021_guiyang/raw_bins/checkm1/*/storage/bin_stats.analyze.tsv > /project/def-careg421/ruizhang/2021_guiyang/raw_bins/checkm1/combined_raw_bin_stats.analyze.tsv
cat /project/def-careg421/ruizhang/2021_guiyang/raw_bins/checkm1/*/storage/bin_stats_ext.tsv > /project/def-careg421/ruizhang/2021_guiyang/raw_bins/checkm1/combined_raw_bin_stats_ext.tsv
```

### 1.2 CheckM2

```bash
# Install CheckM2 (v1.0.1)
module load StdEnv/2020
module load python/3.8
cd /project/def-careg421/ruizhang/software
virtualenv --no-download checkm2-1.0.1-ENV
source /project/def-careg421/ruizhang/software/checkm2-1.0.1-ENV/bin/activate
pip install --upgrade pip
pip install tensorflow==2.5.0 --no-index
pip install CheckM2==1.0.1

# Download CheckM2 database
cd /project/def-careg421/ruizhang/software/checkm2/bin
./checkm2 database --download --path /project/def-careg421/ruizhang/software/checkm2/checkm2_db

# Run CheckM2 (example for HX_surface)
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --gpus-per-node=p100l:4
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=24
#SBATCH --mem=0
#SBATCH --time=24:00:0
#SBATCH -J=gy2021_checkm2-hx-surf-GPU.sh

module load StdEnv/2020
module load python/3.8
module load diamond/2.1.6
module load prodigal/2.6.3
source /project/def-careg421/ruizhang/software/checkm2-1.0.1-ENV/bin/activate
export CHECKM2DB="/project/def-careg421/ruizhang/software/checkm2/checkm2_db/CheckM2_database/uniref100.KO.1.dmnd"
db=/project/def-careg421/ruizhang/software/checkm2/checkm2_db/CheckM2_database/uniref100.KO.1.dmnd
cd /project/def-careg421/ruizhang/software/checkm2/bin

site="HX_surface"
input=/project/def-careg421/ruizhang/2021_guiyang/raw_bins/dastool/${site}/bins
output=/project/def-careg421/ruizhang/2021_guiyang/raw_bins/checkm2/${site}

./checkm2 predict \
--threads 24 \
--input $input \
--output-directory $output \
--database_path $db \
-x fasta \
--force

# Combine CheckM2 results
cat /project/def-careg421/ruizhang/2021_guiyang/raw_bins/checkm2/*/quality_report.tsv > /project/def-careg421/ruizhang/2021_guiyang/raw_bins/checkm2/combined_raw_bins_quality_report.tsv
```

## 2. Bin Refinement with Anvio-refine

### 2.1 Creating Anvio-contigs Database

```bash
#!/bin/bash
#SBATCH -c 32
#SBATCH --mem=125G
#SBATCH -t 3:00:00
#SBATCH --array=0-5
#SBATCH -J=gy21_gen_contigs_db.sh

module load StdEnv/2020
module load python/3.10.2
module load prodigal/2.6.3
module load hmmer/3.2.1
module load diamond
source /project/def-careg421/ruizhang/software/anvio-7.1/bin/activate

cd ~
names=($(cat gy2021_coassembly.txt))
echo ${names[${SLURM_ARRAY_TASK_ID}]}
contigdb=$SCRATCH/gy21_anvio_tmp_folder/contigs_db
file=/project/def-careg421/ruizhang/2021_guiyang/megahit/co-assembly
mkdir -p $contigdb

anvi-gen-contigs-database \
-f $file/${names[${SLURM_ARRAY_TASK_ID}]}/final.contigs.reformatted.fa \
-o $contigdb/${names[${SLURM_ARRAY_TASK_ID}]}.db \
-n ''${names[${SLURM_ARRAY_TASK_ID}]}'' \
-T 32
```

### 2.2 Run Anvi-scg-taxonomy on Contigs Database

```bash
scg_db=/project/def-careg421/ruizhang/software/anvio-7.1/scg_db
anvi-setup-scg-taxonomy \
--scgs-taxonomy-data-dir $scg_db --reset

#!/bin/bash
#SBATCH -c 32
#SBATCH --mem=125G
#SBATCH -t 6:00:0
#SBATCH --array=0-5
#SBATCH -J=gy21_anvi-run-scg-contigs.sh

module load StdEnv/2020
module load python/3.10.2
module load prodigal/2.6.3
module load hmmer/3.2.1
module load diamond
source /project/def-careg421/ruizhang/software/anvio-7.1/bin/activate

cd ~
names=($(cat gy2021_coassembly.txt))
echo ${names[${SLURM_ARRAY_TASK_ID}]}
scg_db=/project/def-careg421/ruizhang/software/anvio-7.1/scg_db
contigdb=$SCRATCH/gy21_anvio_tmp_folder/contigs_db

anvi-run-hmms \
-c $contigdb/${names[${SLURM_ARRAY_TASK_ID}]}.db \
--num-thread 32

anvi-run-scg-taxonomy \
-c $contigdb/${names[${SLURM_ARRAY_TASK_ID}]}.db \
--scgs-taxonomy-data-dir $scg_db -T 32
```

### 2.3 Creating Anvio Profile Database

```bash
#!/bin/bash
#SBATCH -c 32
#SBATCH --mem=250G
#SBATCH -t 2:00:0
#SBATCH --array=0-13
#SBATCH -J=gy21_HX-sur_anvi_profile_db.sh

module load StdEnv/2020
module load python/3.10.2
module load prodigal/2.6.3
module load hmmer/3.2.1
module load diamond
source /project/def-careg421/ruizhang/software/anvio-7.1/bin/activate

PROJECT=/project/def-careg421/ruizhang
SCRIPTS=/project/def-careg421/ruizhang/2021_guiyang/scripts
cd ~
names=($(cat gy2021_HX.txt))
echo ${names[${SLURM_ARRAY_TASK_ID}]}
site='HX_surface'
bam=$SCRATCH/gy2021/mapping_coassembly/${site}
contigdb=$SCRATCH/gy21_anvio_tmp_folder/contigs_db
profiledb=$SCRATCH/gy21_anvio_tmp_folder/profile_db/${site}
mkdir -p $profiledb

anvi-profile \
--min-contig-length 1000 \
-i $bam/${names[${SLURM_ARRAY_TASK_ID}]}_fixed.bam \
-c $contigdb/${site}.db \
--output-dir $profiledb/${names[${SLURM_ARRAY_TASK_ID}]}_profiledb \
--num-threads 32 \
--sample-name gy21_${site}_${names[${SLURM_ARRAY_TASK_ID}]}_profileDB \
--skip-hierarchical-clustering \
--write-buffer-size 10000 -W
```

### 2.4 Anvi-merge

```bash
#!/bin/bash
#SBATCH -c 32
#SBATCH --mem=125G
#SBATCH -t 2:00:0
#SBATCH -J=gy21_anvi-merge-HX-sur-run2.sh

module load StdEnv/2020
module load python/3.10.2
module load prodigal/2.6.3
module load hmmer/3.2.1
module load diamond
source /project/def-careg421/ruizhang/software/anvio-7.1/bin/activate

site='HX_surface'
contigdb=$SCRATCH/gy21_anvio_tmp_folder/contigs_db
profiledb=$SCRATCH/gy21_anvio_tmp_folder/profile_db/${site}
merged_db=$SCRATCH/gy21_anvio_tmp_folder/merged_db
mkdir $merged_db

anvi-merge \
$profiledb/*_profiledb/PROFILE.db \
-o $merged_db/${site}_merged_db \
-c $contigdb/${site}.db \
--enforce-hierarchical-clustering -W
```

### 2.5 Import Collections

```bash
module load StdEnv/2020
module load python/3.10.2
module load prodigal/2.6.3
module load hmmer/3.2.1
module load diamond
source /project/def-careg421/ruizhang/software/anvio-7.1/bin/activate

list='HX_surface'
for site in $list
do
bin=/project/def-careg421/ruizhang/2021_guiyang/raw_bins/dastool/${site}
contigdb=$SCRATCH/gy21_anvio_tmp_folder/contigs_db
merged_db=$SCRATCH/gy21_anvio_tmp_folder/merged_db

anvi-import-collection \
$bin/dastool_run2_DASTool_contig2bin-rename.tsv \
-p $merged_db/${site}_merged_db/PROFILE.db \
-c $contigdb/${site}.db \
-C "${site}_Dastool_bins" \
--contigs-mode
done
```

### 2.6 Refinement Step

```bash
# Login to the cluster 
alias cedar="ssh -L 8090:localhost:8090 ruizhang@cedar.computecanada.ca | tee /dev/tty | python3 ~/.ssh/run_webbrowser.py"
cedar

# Assign port number
[[ "$(whoami)" = "ruizhang" ]] && export ANVIO_PORT=8090
echo $ANVIO_PORT

# Run refinement
module load StdEnv/2020
module load python/3.10.2
module load prodigal/2.6.3
module load hmmer/3.2.1
module load diamond
source /project/def-careg421/ruizhang/software/anvio-7.1/bin/activate

site='SK_surface'
bin='metabat2_SK_surface_bin_27'
contigdb=$SCRATCH/gy21_anvio_tmp_folder/contigs_db
merged_db=$SCRATCH/gy21_anvio_tmp_folder/merged_db

anvi-refine \
-c $contigdb/${site}.db \
-p $merged_db/${site}_merged_db/PROFILE.db \
-C "${site}_Dastool_bins" \
-b ${bin} \
--skip-init-functions \
--skip-news
```

## 3. Dereplication

```bash
#!/bin/bash
#SBATCH -c 32
#SBATCH --mem=125G
#SBATCH -t 14:00:0
#SBATCH -J=gy21_bins_drep-raw-bins.sh

module load StdEnv/2020 gcc/9.3.0
module load python/3.11.2
module load mummer
module load mash
module load prodigal
module load fastani
module load centrifuge
module load hmmer/3.3.2
source /project/def-careg421/ruizhang/software/drep-ENV/bin/activate
export PATH="/project/def-careg421/ruizhang/software/ANIcalculator_v1:$PATH"
export PATH="/project/def-careg421/ruizhang/software/qsimscan:$PATH"
export CHECKM_DATA_PATH=/project/def-careg421/ruizhang/software/checkm1_db
export PATH="/project/def-careg421/ruizhang/software/checkm1_db:$PATH"

PROJECT=/project/def-careg421/ruizhang
input=$PROJECT/2021_guiyang/final_bins
output=$PROJECT/2021_guiyang/final_bins_drep
mkdir $output

# Should include the full name of the genome files including .fasta in the genomeInfo file
# Providing checkm2 genome info
dRep dereplicate -p 32 $output -g $input/*.fasta \
--genomeInfo $input/checkm2/finalbin_genome_info_checkm2.csv \
--S_ani 0.98
```

## 4. GTDB Classification

### 4.1 Installation

```bash
module load StdEnv/2020 gcc/9.3.0
module load python/3.11
module load prodigal/2.6.3
module load hmmer/3.3.2
module load pplacer/1.1.alpha19
module load fastani/1.32
module load fasttree/2.1.11
module load mash/2.3

cd /project/def-careg421/ruizhang/software/
virtualenv --no-download GTDB-tk
source /project/def-careg421/ruizhang/software/GTDB-tk/bin/activate
pip install gtdbtk==2.3.0

# Get reference data R214
wget https://data.gtdb.ecogenomic.org/releases/release214/214.0/auxillary_files/gtdbtk_r214_data.tar.gz
gzip -d *.fasta.gz
```

### 4.2 Run GTDB-tk on Final Bins

```bash
#!/bin/bash
#SBATCH -c 32
#SBATCH --mem=125G
#SBATCH -t 1-00:00:0
#SBATCH -J=gy2021_gtdb-final-bins.sh

module load StdEnv/2020 gcc/9.3.0
module load python/3.11
module load prodigal/2.6.3
module load hmmer/3.3.2
module load pplacer/1.1.alpha19
module load fastani/1.32
module load fasttree/2.1.11
module load mash/2.3
source /project/def-careg421/ruizhang/software/GTDB-tk/bin/activate

# Set environment variable $GTDBTK_DATA_PATH
export GTDBTK_DATA_PATH=/project/def-careg421/ruizhang/software/release214

site='final_bins'
PROJECT=/project/def-careg421/ruizhang
mkdir -p $SCRATCH/gy2021/gtdb-tk/${site}/identify_rs214/
mkdir -p $SCRATCH/gy2021/gtdb-tk/${site}/align_rs214/
mkdir -p $SCRATCH/gy2021/gtdb-tk/${site}/classify_rs214/
genome=$PROJECT/2021_guiyang/final_bins
input=$SCRATCH/gy2021/gtdb-tk/${site}

gtdbtk identify \
--genome_dir $genome \
--out_dir $input/identify_rs214 \
--extension fasta --cpus 32

gtdbtk align \
--identify_dir $input/identify_rs214 \
--out_dir $input/align_rs214 --cpus 32

gtdbtk classify \
--genome_dir $genome \
--align_dir $input/align_rs214 \
--out_dir $input/classify_rs214 \
--scratch_dir $input/classify_rs214 \
--mash_db $SCRATCH/mash_db \
-x fasta \
--cpus 32 \
--pplacer_cpus 1
```

## 5. MAGs Coverage Calculation

```bash
#!/bin/bash
#SBATCH -c 32
#SBATCH --mem=125G
#SBATCH -t 12:00:0
#SBATCH -J=gy21_HX_coverM_RA.sh

module load StdEnv/2020
module load samtools
module load minimap2
module load bwa-mem2
module load fastani
export PATH="/project/def-careg421/ruizhang/software/coverm-0.6.1:$PATH"
export PATH="/project/def-careg421/ruizhang/software/dashing-binaries-1.0.3/linux:$PATH"

site='HX'
reads=/project/def-careg421/ruizhang/2021_guiyang/fastp/dna
genomes=/project/def-careg421/ruizhang/2021_guiyang/20240731_drep/output_bins/${site}/dereplicated_genomes

list="${reads}/gy2021_HX_sur_t1.R1.fq.gz ${reads}/gy2021_HX_sur_t1.R2.fq.gz ${reads}/gy2021_HX_sur_t2.R1.fq.gz ${reads}/gy2021_HX_sur_t2.R2.fq.gz ${reads}/gy2021_HX_sur_t3.R1.fq.gz ${reads}/gy2021_HX_sur_t3.R2.fq.gz ${reads}/gy2021_HX_sur_t4.R1.fq.gz ${reads}/gy2021_HX_sur_t4.R2.fq.gz ${reads}/gy2021_HX_sur_t5.R1.fq.gz ${reads}/gy2021_HX_sur_t5.R2.fq.gz ${reads}/gy2021_HX_sur_t6.R1.fq.gz ${reads}/gy2021_HX_sur_t6.R2.fq.gz ${reads}/gy2021_HX_sur_t7.R1.fq.gz ${reads}/gy2021_HX_sur_t7.R2.fq.gz ${reads}/gy2021_HX_sur_t8.R1.fq.gz ${reads}/gy2021_HX_sur_t8.R2.fq.gz ${reads}/gy2021_HX_sur_t9.R1.fq.gz ${reads}/gy2021_HX_sur_t9.R2.fq.gz ${reads}/gy2021_HX_sur_t10.R1.fq.gz ${reads}/gy2021_HX_sur_t10.R2.fq.gz ${reads}/gy2021_HX_sur_t11.R1.fq.gz ${reads}/gy2021_HX_sur_t11.R2.fq.gz ${reads}/gy2021_HX_rhi_t2.R1.fq.gz ${reads}/gy2021_HX_rhi_t2.R2.fq.gz ${reads}/gy2021_HX_rhi_t3.R1.fq.gz ${reads}/gy2021_HX_rhi_t3.R2.fq.gz ${reads}/gy2021_HX_rhi_t4.R1.fq.gz ${reads}/gy2021_HX_rhi_t4.R2.fq.gz ${reads}/gy2021_HX_rhi_t5.R1.fq.gz ${reads}/gy2021_HX_rhi_t5.R2.fq.gz ${reads}/gy2021_HX_rhi_t6.R1.fq.gz ${reads}/gy2021_HX_rhi_t6.R2.fq.gz ${reads}/gy2021_HX_rhi_t7.R1.fq.gz ${reads}/gy2021_HX_rhi_t7.R2.fq.gz"

# Don't need to create the output in advance
output=/project/def-careg421/ruizhang/2021_guiyang/coverM

coverm genome \
--coupled $list \
--genome-fasta-directory $genomes \
--genome-fasta-extension fasta \
--output-file $output/gy21_${site}_MAGs_coverM_relative_abundance.txt \
--methods relative_abundance \
-t 32
```
