# MAGs Functional Annotation Protocol

## 1. MAGs Methanogen and Methanotroph Identification

### 1.1 Running Prodigal to Identify ORFs

```bash
module load prodigal/2.6.3

MAGs=/project/def-careg421/ruizhang/2021_guiyang/final_bins_20230720
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/MAGs-annotation

for file in $MAGs/*; do
  # Extract just the base filename without the path
  basefile=$(basename "$file")
    
  # Now strip the extension from the base filename
  filename="${basefile%.*}"
  
  # Print or use the filename without extension
  echo "${filename}"
  
  # Run prodigal
  prodigal \
  -i $MAGs/${filename}.fasta \
  -a $annotation/${filename}_ORF.faa \
  -p single
  
  # Adding file names to the beginning of each header  
  sed -i 's/^>/&'${filename}' /' $annotation/${filename}_ORF.faa
done
```



### 1.2 Reformatting MAGs Annotation Files

```bash
MAGs=/project/def-careg421/ruizhang/2021_guiyang/final_bins_20230720
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/MAGs-annotation
annotation_reformat=$SCRATCH/tmp/MAGs
mkdir $annotation_reformat

while read -r MAG 
do
awk '{sub(/ /, "_", $0); print}' $annotation/${MAG}_ORF.faa > $annotation_reformat/${MAG}_ORF.faa
echo $MAG
done < ~/gy21_MAGs_list.txt
```

### 1.3 Running HMM Searches for mcrA, pmoA and mmoX

```bash
module load StdEnv/2020
module load hmmer
module load seqtk/1.3

MAGs=/project/def-careg421/ruizhang/2021_guiyang/final_bins_20230720
annotation_reformat=$SCRATCH/tmp/MAGs
hmm_out=$SCRATCH/tmp/hmm_out

# For pmoA
hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm/mcrA_PFAM.hmm
gene='mcrA_PFAM'

# For pmoA
hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm/pmoA_graftM.hmm
gene='pmoA_graftM'

# For mmoX
# hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm/mmoX_graftM.hmm
# gene='mmoX_graftM'

while read -r MAG
do
# run HMM
hmmsearch -E 0.00001 -o ${hmm_out}/${MAG}_${gene}.txt --tblout ${hmm_out}/${MAG}_${gene}_hmmer.out $hmm $annotation_reformat/${MAG}_ORF.faa

# extracting sequences that passed the filter
grep -v '^#' ${hmm_out}/${MAG}_${gene}_hmmer.out | awk {'{print $1}'} | sort > ${hmm_out}/${MAG}_${gene}_geneid.txt

seqtk subseq $annotation_reformat/${MAG}_ORF.faa ${hmm_out}/${MAG}_${gene}_geneid.txt | awk {'{gsub(/*$/,""); print}'} > ${hmm_out}/${MAG}_${gene}_proteins.faa

echo $MAG

done < ~/gy21_MAGs_list.txt

# combine files
cd $hmm_out
cat *_${gene}_geneid.txt > gy21_MAGs_${gene}_IDs.txt
cat *_${gene}_proteins.faa > gy21_MAGs_${gene}_proteins.faa
```

## 2. Methylator Identification

### 2.1 Extracting hgcA from MAGs

```bash
MAGs=/project/def-careg421/ruizhang/2021_guiyang/final_bins_20230720
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/MAGs-annotation
hgcA=/project/def-careg421/ruizhang/software/anvio-7.1/hmm/anvio_custom_hmm_hgcA
hmm_out=$SCRATCH/gy2021/hmm_MAGs

for file in $MAGs/*; do
  # Extract just the base filename without the path
  basefile=$(basename "$file")
    
  # Now strip the extension from the base filename
  filename="${basefile%.*}"
  
  # Print or use the filename without extension
  echo "${filename}"

  # use a cut of of 1e-50 to filter out low confidence hgcA hits
  module load StdEnv/2023
  module load hmmer/3.4 

  hmmsearch -E 1e-50 -o ${hmm_out}/${filename}_hgcA.txt --tblout ${hmm_out}/${filename}_hgcA_hmmer.out $hgcA/genes.hmm $annotation/${filename}_ORF.faa

  # extracting the sequences that passed the filter
  grep -v '^#' ${hmm_out}/${filename}_hgcA_hmmer.out | awk {'{print $1}'} | sort > ${hmm_out}/${filename}_hgcA_geneid.txt

  module load nixpkgs/16.09 gcc/7.3.0
  module load seqtk/1.3

  seqtk subseq $annotation/${filename}_ORF.faa ${hmm_out}/${filename}_hgcA_geneid.txt | awk {'{gsub(/*$/,""); print}'} > ${hmm_out}/${filename}_hgcA_proteins.faa

  # retain sequences that contain the cap helix motif
  awk '/^>/ {printf("%s%s\t",(N>0?"\n":""),$0);N++;next;} {printf("%s",$0);} END {printf("\n");}' ${hmm_out}/${filename}_hgcA_proteins.faa |\
  awk -F '\t' '{if(index($2,"NVWCAAGK")!=0) printf("%s\n%s\n",$1,$2);}' > ${hmm_out}/${filename}_hgcA_proteins_filtered1.fasta

  awk '/^>/ {printf("%s%s\t",(N>0?"\n":""),$0);N++;next;} {printf("%s",$0);} END {printf("\n");}' ${hmm_out}/${filename}_hgcA_proteins.faa |\
  awk -F '\t' '{if(index($2,"NVWCASGK")!=0) printf("%s\n%s\n",$1,$2);}' >> ${hmm_out}/${filename}_hgcA_proteins_filtered1.fasta

  awk '/^>/ {printf("%s%s\t",(N>0?"\n":""),$0);N++;next;} {printf("%s",$0);} END {printf("\n");}' ${hmm_out}/${filename}_hgcA_proteins.faa |\
  awk -F '\t' '{if(index($2,"NVWCAGGK")!=0) printf("%s\n%s\n",$1,$2);}' >> ${hmm_out}/${filename}_hgcA_proteins_filtered1.fasta

  awk '/^>/ {printf("%s%s\t",(N>0?"\n":""),$0);N++;next;} {printf("%s",$0);} END {printf("\n");}' ${hmm_out}/${filename}_hgcA_proteins.faa |\
  awk -F '\t' '{if(index($2,"NIWCAAGK")!=0) printf("%s\n%s\n",$1,$2);}' >> ${hmm_out}/${filename}_hgcA_proteins_filtered1.fasta

  awk '/^>/ {printf("%s%s\t",(N>0?"\n":""),$0);N++;next;} {printf("%s",$0);} END {printf("\n");}' ${hmm_out}/${filename}_hgcA_proteins.faa |\
  awk -F '\t' '{if(index($2,"NIWCAGGK")!=0) printf("%s\n%s\n",$1,$2);}' >> ${hmm_out}/${filename}_hgcA_proteins_filtered1.fasta

  awk '/^>/ {printf("%s%s\t",(N>0?"\n":""),$0);N++;next;} {printf("%s",$0);} END {printf("\n");}' ${hmm_out}/${filename}_hgcA_proteins.faa |\
  awk -F '\t' '{if(index($2,"NVWCSAGK")!=0) printf("%s\n%s\n",$1,$2);}' >> ${hmm_out}/${filename}_hgcA_proteins_filtered1.fasta

  echo "${filename} completed"
done

# merge all files
output=/project/def-careg421/ruizhang/2021_guiyang/hgcAB_analysis

cat ${hmm_out}/*_hgcA_proteins_filtered1.fasta > ${output}/gy21_MAGs_hgcA_proteins_filtered1.fasta

# extracting headers and see which MAGs contain hgcA
grep ">" ${output}/gy21_MAGs_hgcA_proteins_filtered1.fasta > ${output}/gy21_MAGs_hgcA_proteins_headers.txt

# extract the files names
awk '{print $1}' ${output}/gy21_MAGs_hgcA_proteins_headers.txt > ${output}/gy21_MAGs_id.txt
sed -i 's/>//g' ${output}/gy21_MAGs_id.txt
```

### 2.2 Extracting hgcB from MAGs

```bash
hgcB=/project/def-careg421/ruizhang/software/anvio-7.1/hmm/anvio_custom_hmm_hgcB

for file in $MAGs/*; do
  # Extract just the base filename without the path
  basefile=$(basename "$file")
    
  # Now strip the extension from the base filename
  filename="${basefile%.*}"
  
  # Print or use the filename without extension
  echo "${filename}"

  # use a cut of of 1e-50 to filter out low confidence hgcA hits
  module load StdEnv/2023
  module load hmmer/3.4 

  hmmsearch -E 1e-30 -o ${hmm_out}/${filename}_hgcB.txt --tblout ${hmm_out}/${filename}_hgcB_hmmer.out $hgcB/genes.hmm $annotation/${filename}_ORF.faa

  # extracting the sequences that passed the filter
  grep -v '^#' ${hmm_out}/${filename}_hgcB_hmmer.out | awk {'{print $1}'} | sort > ${hmm_out}/${filename}_hgcB_geneid.txt

  module load nixpkgs/16.09 gcc/7.3.0
  module load seqtk/1.3

  seqtk subseq $annotation/${filename}_ORF.faa ${hmm_out}/${filename}_hgcB_geneid.txt | awk {'{gsub(/*$/,""); print}'} > ${hmm_out}/${filename}_hgcB_proteins.faa

  # retain sequences that contain the motifs featuring the conserved Cys 
  awk '/^>/ {printf("%s%s\t",(N>0?"\n":""),$0);N++;next;} {printf("%s",$0);} END {printf("\n");}' ${hmm_out}/${filename}_hgcB_proteins.faa |\
  awk -F '\t' '{if(index($2,"CMECGA")!=0) printf("%s\n%s\n",$1,$2);}' > ${hmm_out}/${filename}_hgcB_proteins_filtered1.fasta

  awk '/^>/ {printf("%s%s\t",(N>0?"\n":""),$0);N++;next;} {printf("%s",$0);} END {printf("\n");}' ${hmm_out}/${filename}_hgcB_proteins.faa |\
  awk -F '\t' '{if(index($2,"CIECGA")!=0) printf("%s\n%s\n",$1,$2);}' >> ${hmm_out}/${filename}_hgcB_proteins_filtered1.fasta

  echo "${filename} completed"
done

# merge all files
output=/project/def-careg421/ruizhang/2021_guiyang/hgcAB_analysis
cat ${hmm_out}/*_hgcB_proteins_filtered1.fasta > ${output}/gy21_MAGs_hgcB_proteins_filtered1.fasta

# extracting headers and see which MAGs contain hgcA
grep "^>" ${output}/gy21_MAGs_hgcB_proteins_filtered1.fasta > ${output}/gy21_MAGs_hgcB_proteins_headers.txt

# extract the files names
awk '{print $1}' ${output}/gy21_MAGs_hgcB_proteins_headers.txt > ${output}/gy21_MAGs_id_hgcB.txt
sed -i 's/>//g' ${output}/gy21_MAGs_id_hgcB.txt
```

## 3. Sulfur Reducer Identification

### 3.1 Extracting DsrA from MAGs

```bash
MAGs=/project/def-careg421/ruizhang/2021_guiyang/final_bins_20230720
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/MAGs-annotation
hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm
hmm_out=$SCRATCH/gy2021/hmm_MAGs

gene='DsrA'

for file in $MAGs/*; do
  # Extract just the base filename without the path
  basefile=$(basename "$file")
    
  # Now strip the extension from the base filename
  filename="${basefile%.*}"
  
  # Print or use the filename without extension
  echo "${filename}"

  # use a cut of of 1e-50 to filter out low confidence hgcA hits
  module load StdEnv/2023
  module load hmmer/3.4 

  hmmsearch -T 209.3 -o ${hmm_out}/${filename}_${gene}.txt --tblout ${hmm_out}/${filename}_${gene}_hmmer.out $hmm/TIGR02064.1.hmm $annotation/${filename}_ORF.faa

  # extracting the sequences that passed the filter
  grep -v '^#' ${hmm_out}/${filename}_${gene}_hmmer.out | awk {'{print $1}'} | sort > ${hmm_out}/${filename}_${gene}_geneid.txt

  module load nixpkgs/16.09 gcc/7.3.0
  module load seqtk/1.3

  seqtk subseq $annotation/${filename}_ORF.faa ${hmm_out}/${filename}_${gene}_geneid.txt | awk {'{gsub(/*$/,""); print}'} > ${hmm_out}/${filename}_${gene}_proteins.faa
done

cat ${hmm_out}/*_${gene}_geneid.txt > ${hmm_out}/gy21_MAGs_${gene}_geneid_combined.txt
cat ${hmm_out}/*_${gene}_proteins.faa > ${hmm_out}/gy21_MAGs_${gene}_proteins_combined.fasta
```

### 3.2 Extracting DsrB from MAGs

```bash
MAGs=/project/def-careg421/ruizhang/2021_guiyang/final_bins_20230720
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/MAGs-annotation
hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm
hmm_out=$SCRATCH/gy2021/hmm_MAGs

gene='DsrB'

for file in $MAGs/*; do
  # Extract just the base filename without the path
  basefile=$(basename "$file")
    
  # Now strip the extension from the base filename
  filename="${basefile%.*}"
  
  # Print or use the filename without extension
  echo "${filename}"

  # use a cut of of 1e-50 to filter out low confidence hgcA hits
  module load StdEnv/2023
  module load hmmer/3.4 

  hmmsearch -T 345.1 -o ${hmm_out}/${filename}_${gene}.txt --tblout ${hmm_out}/${filename}_${gene}_hmmer.out $hmm/TIGR02066.1.hmm $annotation/${filename}_ORF.faa

  # extracting the sequences that passed the filter
  grep -v '^#' ${hmm_out}/${filename}_${gene}_hmmer.out | awk {'{print $1}'} | sort > ${hmm_out}/${filename}_${gene}_geneid.txt

  module load nixpkgs/16.09 gcc/7.3.0
  module load seqtk/1.3

  seqtk subseq $annotation/${filename}_ORF.faa ${hmm_out}/${filename}_${gene}_geneid.txt | awk {'{gsub(/*$/,""); print}'} > ${hmm_out}/${filename}_${gene}_proteins.faa
done

cat ${hmm_out}/*_${gene}_geneid.txt > ${hmm_out}/gy21_MAGs_${gene}_geneid_combined.txt
```

### 3.3 Finding MAGs with both dsrA and dsrB

```bash
cd $SCRATCH
sort dsrA.txt -o dsrA_sorted.txt
sort dsrB.txt -o dsrB_sorted.txt
comm -12 dsrA_sorted.txt dsrB_sorted.txt
```

## 4. MerB Encoder Identification

### 4.1 Extracting merB from MAGs

```bash
module load StdEnv/2020
module load hmmer/3.3.2
module load seqtk/1.3

MAGs=/project/def-careg421/ruizhang/2021_guiyang/final_bins_20230720
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/MAGs-annotation
hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm
hmm_out=$SCRATCH/gy2021/hmm_MAGs

gene='MerB'

for file in $MAGs/*; do
  # Extract just the base filename without the path
  basefile=$(basename "$file")
    
  # Now strip the extension from the base filename
  filename="${basefile%.*}"
  
  # Print or use the filename without extension
  echo "${filename}"

  # hmmsearch -E 0.00001 $hmm/PF03243.hmm $annotation/${filename}_ORF.faa > ${hmm_out}/${filename}_${gene}.txt
  sed -n -e '18,/Domain/p' ${hmm_out}/${filename}_${gene}.txt | head -n -3  | sed 's/ \+/|/g' | cut -f10,11 -d"|" > ${hmm_out}/${filename}_${gene}2.txt

  sed -i 's/|/_/g' ${hmm_out}/${filename}_${gene}2.txt

  awk '{print $1 " " $2}' $annotation/${filename}_ORF.faa > $annotation/${filename}_ORF_trim_header.faa

  sed -i 's/ /_/g' $annotation/${filename}_ORF_trim_header.faa

  # get protein sequence
  seqtk subseq $annotation/${filename}_ORF_trim_header.faa ${hmm_out}/${filename}_${gene}2.txt | awk {'{gsub(/*$/,""); print}'} > ${hmm_out}/${filename}_${gene}_protein_seq.fasta
done

# get all sequence
cat ${hmm_out}/*_${gene}_protein_seq.fasta > ${hmm_out}/gy21_MAGs_${gene}_protein_combined.txt

# select for true merB hit
awk '/^>/ {if (NR > 1) printf("\n"); print; next} {printf("%s", $0)} END {printf("\n")}' ${hmm_out}/gy21_MAGs_${gene}_protein_combined.txt > ${hmm_out}/gy21_MAGs_${gene}_protein_combined1.txt

# select lines with specific patterns
awk '/^>/ {header=$0; next} /C...D......................................................................................................................................................................C....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................C/ {print header; print}' ${hmm_out}/gy21_MAGs_${gene}_protein_combined1.txt > ${hmm_out}/gy21_MAGs_${gene}_protein_combined2.txt 

grep ">" ${hmm_out}/gy21_MAGs_${gene}_protein_combined1.txt >  ${hmm_out}/gy21_MAGs_${gene}_seqID.txt
```

### 4.2 BLAST Verification of merB Hits

```bash
#!/bin/bash
#SBATCH -c 32
#SBATCH --mem=125G
#SBATCH -t 6:00:0
#SBATCH -J=blastp-merB-MAGs.sh

module load StdEnv/2020 gcc/9.3.0
module load blast+/2.14.0

cd /project/def-warren21/ruizhang/software/blast_db
export BLASTDB=$BLASTDB:/project/def-warren21/ruizhang/software/blast_db

hmm_out=$SCRATCH/gy2021/hmm_MAGs
gene='MerB'

database=/project/def-warren21/ruizhang/software/blast_db/nr
input=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output
output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/blastp-output

blastp -db $database \
-query ${hmm_out}/gy21_MAGs_${gene}_protein_combined1.txt \
-out $output/gy21_${gene}_MAGs_blastp.txt \
-max_target_seqs 5 \
-outfmt "6 qacc sacc stitle staxids sscinames scomnames evalue bitscore" \
-num_threads 32

# Select the first record of each BLAST sequence hit
awk 'NR % 5 == 1' $output/gy21_${gene}_MAGs_blastp.txt > $output/gy21_${gene}_MAGs_blastp_first_hit.txt

# Only retain those that hit mercury
grep -i "merc" $output/gy21_${gene}_MAGs_blastp_first_hit.txt > $output/gy21_${gene}_MAGs_blastp_first_hit_trimmed.txt
```
