# Gene-centric Analyses

## 1. FeatureCount

```bash
#!/bin/bash
#SBATCH --time=10:30:00
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=32
#SBATCH --mem=64G
#SBATCH --job-name=FC_gene_count_array
#SBATCH --array=0-47

module load StdEnv/2020
module load gcc/9.3.0
module load anvio/7.1
module load prodigal/2.6.3
module load samtools/1.17
module load bwa/0.7.17
module load subread/2.0.3

input=/project/def-careg421/ruizhang/2021_guiyang/individual-assembly
assembly=/project/def-careg421/ruizhang/2021_guiyang/individual-assembly-scaffolds
output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-featureCounts
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-annotation

# Read sample names from file
mapfile -t sample_array < ~/gy21_dna_names.txt

# Get sample name
sample_name=${sample_array[$SLURM_ARRAY_TASK_ID]}

# Step1: Rename scaffolds
cd $input
mkdir -p $SLURM_TMPDIR/tmp
anvi-script-reformat-fasta $assembly/${sample_name}_scaffolds.fasta -o $SLURM_TMPDIR/tmp/${sample_name}_scaffolds_revised.fasta --prefix $sample_name --simplify-names

# Step2: Predict genes
cd $SLURM_TMPDIR/tmp
prodigal -i ${sample_name}_scaffolds_revised.fasta -a ${sample_name}_proteins.faa -d ${sample_name}_genes.fna -f gff -o ${sample_name}_prodigal.gff
cp ${sample_name}_proteins.faa ${sample_name}_genes.fna ${sample_name}_prodigal.gff $annotation/

# Step3: BWA mapping to get .sam file
cd $SLURM_TMPDIR/tmp
bwa index ${sample_name}_scaffolds_revised.fasta
rm -rf *.pac *.bwt *.sa *.amb *.ann

# Example for HX site
bwa mem -t 32 ${sample_name}_scaffolds_revised.fasta \
/project/def-careg421/ruizhang/2021_guiyang/trimmed-reads/HX_surface_S1_R1_trimmed.fastq.gz \
/project/def-careg421/ruizhang/2021_guiyang/trimmed-reads/HX_surface_S1_R2_trimmed.fastq.gz > ${sample_name}_HX_surface.sam

cp ${sample_name}_*.sam $annotation/

# Step4: turn .sam into .bam file
cd $annotation
samtools view -@ 32 -bS ${sample_name}_HX_surface.sam | samtools sort -@ 32 -o ${sample_name}_HX_surface.bam -
samtools index ${sample_name}_HX_surface.bam

# Step5: FeatureCounts
# -g is the name of attribute which is used to group features (exons, etc.) into meta-features (genes)
cd $annotation
featureCounts \
-T 32 \
-t CDS \
-g ID \
-a ${sample_name}_prodigal.gff \
-o $output/${sample_name}_HX_surface.count \
${sample_name}_HX_surface.bam

echo "$sample_name feature count completed"

# Create a log file to track completion
echo "$sample_name" >> ~/featurecount_completed.txt
```

## 2. Function-annotation with HMM

```bash
#!/bin/bash
#SBATCH --time=20:30:00
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=32
#SBATCH --mem=64G
#SBATCH --job-name=FA_HMM_array
#SBATCH --array=0-47

module load StdEnv/2023
module load hmmer/3.4

# Read sample names from file
mapfile -t sample_array < ~/gy21_dna_names.txt

# Get sample name
sample_name=${sample_array[$SLURM_ARRAY_TASK_ID]}

# Change the following parameter
gene='mmoX_graftM'
hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm/${gene}.hmm

# File paths
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-annotation
featureCounts=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-featureCounts
output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output

# Step 1 - run hmmsearch to determine gene id and their coverage
tmp=$SCRATCH/gy2021/gene_coverage/${sample_name}_tmp
mkdir -p $tmp

hmmsearch -E 0.00001 $hmm $annotation/${sample_name}_proteins.faa > $output/${sample_name}_${gene}.txt
sed -n -e '18,/Domain/p' $output/${sample_name}_${gene}.txt | head -n -3 | sed 's/ \+/|/g' | cut -f10 -d"|" > $tmp/${sample_name}_${gene}2.txt

# Step 2 - extract featureCounts for the selected gene ids
cd $tmp

cp $annotation/${sample_name}_proteins.faa .
cp $featureCounts/${sample_name}_HX_surface.count $tmp/${sample_name}_gene_count.csv

# Get the sequence ID of the selected gene from hmmsearch result
while read -r gene_id
do
  # Remove * from the output
  gene_id=$(echo "$gene_id" | sed 's/\*//g')
  
  # Search for gene_id in the featureCounts file, extract relevant info
  grep "${gene_id}" $tmp/${sample_name}_gene_count.csv >> $tmp/${sample_name}_${gene}_count.csv
done < $tmp/${sample_name}_${gene}2.txt

# Step 3 - normalize with gene length
module load StdEnv/2020
module load seqtk/1.3

# Calculate gene length
cd $tmp

# Get the protein sequence for each HMM hit
seqtk subseq ${sample_name}_proteins.faa $tmp/${sample_name}_${gene}2.txt | awk '{gsub(/*$/,""); print}' > ${sample_name}_${gene}_proteins.faa

# Calculate alignment length to the reference HMM (a better way to calculate coverage)
sed -n -e '18,/Domain/p' $output/${sample_name}_${gene}.txt | head -n -3 | sed 's/ \+/|/g' | cut -f9,10 -d"|" > ${sample_name}_${gene}_overlap.txt

# Process the file to a format include only the header
awk -F' ' '{print $1}' ${sample_name}_${gene}_proteins.faa > ${sample_name}_${gene}_proteins2.faa

# Add site name to the header
sed "s/^/${sample_name}_/" ${sample_name}_${gene}_proteins2.faa > ${sample_name}_${gene}_proteins3.faa

# Transfer result to the output folder
cp ${sample_name}_${gene}_proteins3.faa $output/

# Extract read count of the gene
if [ -s "$tmp/${sample_name}_${gene}_count.csv" ]; then
  cut -f1,7 $tmp/${sample_name}_${gene}_count.csv > $tmp/${sample_name}_${gene}_raw_count.csv
  sed 's/\t/,/g' $tmp/${sample_name}_${gene}_raw_count.csv > $tmp/${sample_name}_${gene}_raw_count_comma.csv
  cp $tmp/${sample_name}_${gene}_raw_count_comma.csv $output/
  echo "$sample_name ${gene} identification completed with hits"
else
  touch $tmp/${sample_name}_${gene}_raw_count_comma.csv
  cp $tmp/${sample_name}_${gene}_raw_count_comma.csv $output/
  echo "$sample_name ${gene} identification completed with no hits"
fi

# Create a log file to track completion
echo "$sample_name" >> ~/HMM_completed_${gene}.txt
```



## 3. Blast

### 3.1 mcrA

```bash
# Before running BLAST, we need to run hmmsearch to identify amino acid sequences of interest
module load StdEnv/2023
module load hmmer/3.4

# Change the following parameter
gene='mcrA_Pfam'
hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm/${gene}.hmm

# File paths
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-annotation
featureCounts=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-featureCounts
output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output

# Running BLAST using the NR database
module load StdEnv/2020 gcc/9.3.0
module load blast+/2.14.0
cd /project/def-warren21/ruizhang/software/blast_db
export BLASTDB=$BLASTDB:/project/def-warren21/ruizhang/software/blast_db
database=/project/def-warren21/ruizhang/software/blast_db/nr
input=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output
output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/blastp-output

blastp -db $database \
-query $input/AA_gy21_${gene}_combined_sequence.faa \
-out $output/gy21_${gene}_blastp.txt \
-max_target_seqs 5 \
-outfmt "6 qacc sacc stitle staxids sscinames scomnames evalue bitscore" \
-num_threads 32

# Select the first record of each BLAST sequence hit
awk 'NR % 5 == 1' $output/gy21_${gene}_blastp.txt > $output/gy21_${gene}_blastp_first_hit.txt

# Examining sequence title, only retain those that hit methane monooxygenase, or with taxonomy containing methylo- prefix
grep -i "meth" $output/gy21_${gene}_blastp_first_hit.txt > $output/gy21_${gene}_blastp_first_hit_trimmed.txt

# Select specific columns of the BLAST output file
cut -f 1,4 $output/gy21_${gene}_blastp_first_hit_trimmed.txt > $output/gy21_${gene}_blastp_formatted.txt

# Some sequence have tax ID like this 143111;143114, remove the part after the semi colon
cut -d';' -f1 $output/gy21_${gene}_blastp_formatted.txt > $output/gy21_${gene}_blastp_formatted1.txt

# Split the output files into different files according to sites
while read -r INPUT
do
  grep "${INPUT}_" $output/gy21_${gene}_blastp_formatted1.txt > $output/gy21_${INPUT}_${gene}_blastp.txt
done < ~/gy2021_all_samples.txt
```

### 3.2 pmoA/mmoX

```bash
# Before running BLAST, we need to run hmmsearch to identify amino acid sequences of interest
module load StdEnv/2023
module load hmmer/3.4

# Change the following parameter for pmoA
gene='pmoA_graftM'
hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm/${gene}.hmm

# File paths
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-annotation
featureCounts=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-featureCounts
output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output

# Example for HX site
while read -r INPUT
do
  # Step 1 - run hmmsearch to determine gene id and their coverage
  tmp=$SCRATCH/gy2021/gene_coverage/${INPUT}_tmp
  mkdir -p $tmp
  
  hmmsearch -E 0.00001 $hmm $annotation/${INPUT}_proteins.faa > $output/${INPUT}_${gene}.txt
  sed -n -e '18,/Domain/p' $output/${INPUT}_${gene}.txt | head -n -3  | sed 's/ \+/|/g' | cut -f10 -d"|" > $tmp/${INPUT}_${gene}2.txt
  echo "$INPUT completed"
done < ~/gy2021_all_samples.txt

# Running BLAST using the NR database
module load StdEnv/2020 gcc/9.3.0
module load blast+/2.14.0
cd /project/def-warren21/ruizhang/software/blast_db
export BLASTDB=$BLASTDB:/project/def-warren21/ruizhang/software/blast_db
database=/project/def-warren21/ruizhang/software/blast_db/nr
input=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output
output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/blastp-output

blastp -db $database \
-query $input/AA_gy21_${gene}_combined_sequence.faa \
-out $output/gy21_${gene}_blastp.txt \
-max_target_seqs 5 \
-outfmt "6 qacc sacc stitle staxids sscinames scomnames evalue bitscore" \
-num_threads 32

# Select the first record of each BLAST sequence hit
awk 'NR % 5 == 1' $output/gy21_${gene}_blastp.txt > $output/gy21_${gene}_blastp_first_hit.txt

# Examining sequence title first before proceeding, only retain those that hit monooxygenase
grep -i "methane" $output/gy21_${gene}_blastp_first_hit.txt > $output/gy21_${gene}_blastp_first_hit_trimmed.txt

# Select specific columns of the BLAST output file
cut -f 1,4 $output/gy21_${gene}_blastp_first_hit_trimmed.txt > $output/gy21_${gene}_blastp_formatted.txt

# Some sequence have tax ID like this 143111;143114, which are pretty much the same taxonomy, remove the part after the semi colon
cut -d';' -f1 $output/gy21_${gene}_blastp_formatted.txt > $output/gy21_${gene}_blastp_formatted1.txt

# Split the output files into different files according to sites
while read -r INPUT
do
  grep "${INPUT}_" $output/gy21_${gene}_blastp_formatted1.txt > $output/gy21_${INPUT}_${gene}_blastp.txt
done < ~/gy2021_all_samples.txt

# Similar process can be applied for mmoX gene by changing the gene variable
```

### 3.3 merB

```bash
# Before running BLAST, we need to run hmmsearch to identify amino acid sequences of interest
module load StdEnv/2023
module load hmmer/3.4

# Change the following parameter
gene='merB'
hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm/${gene}.hmm

# File paths
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-annotation
featureCounts=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-featureCounts
output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output

while read -r INPUT
do
  # Step 1 - run hmmsearch to determine gene id and their coverage
  tmp=$SCRATCH/gy2021/gene_coverage/${INPUT}_tmp
  mkdir -p $tmp
  
  hmmsearch -E 0.00001 $hmm $annotation/${INPUT}_proteins.faa > $output/${INPUT}_${gene}.txt
  sed -n -e '18,/Domain/p' $output/${INPUT}_${gene}.txt | head -n -3  | sed 's/ \+/|/g' | cut -f10 -d"|" > $tmp/${INPUT}_${gene}2.txt
  echo "$INPUT completed"
done < ~/gy2021_all_samples.txt

# Get INPUT sequence for blast
module load StdEnv/2020
module load seqtk/1.3
while read -r INPUT
do
  featureCounts=$SCRATCH/gy2021/individual-assembly-featureCounts
  output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output
  tmp=$SCRATCH/gy2021/gene_coverage/${INPUT}_tmp
  cd $tmp
  
  # Get the protein sequence for each HMM hit
  seqtk subseq ${INPUT}_proteins.faa $tmp/${INPUT}_${gene}2.txt | awk {'{gsub(/*$/,""); print}'} > ${INPUT}_${gene}_proteins.faa
  
  # Process the file to include only the desired header
  awk -F' ' '{print $1}' ${INPUT}_${gene}_proteins.faa > ${INPUT}_${gene}_proteins2.faa
  echo "$INPUT completed"
done < ~/gy2021_all_samples.txt

# Combine results
cd $SCRATCH/gy2021/gene_coverage/
cat *_tmp/*_${gene}_proteins2.faa > $output/AA_gy21_${gene}_combined_sequence.faa

# Running BLAST using the NR database
module load StdEnv/2020 gcc/9.3.0
module load blast+/2.14.0
cd /project/def-warren21/ruizhang/software/blast_db
export BLASTDB=$BLASTDB:/project/def-warren21/ruizhang/software/blast_db
database=/project/def-warren21/ruizhang/software/blast_db/nr
input=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output
output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/blastp-output

blastp -db $database \
-query $input/AA_gy21_${gene}_combined_sequence1.faa \
-out $output/gy21_${gene}_blastp.txt \
-max_target_seqs 5 \
-outfmt "6 qacc sacc stitle staxids sscinames scomnames evalue bitscore" \
-num_threads 32
```

### 4.4 hgcA

```bash
# Before running BLAST, we need to run hmmsearch to identify amino acid sequences of interest
module load StdEnv/2023
module load hmmer/3.4

# Change the following parameter
gene='hgcA'
hmm=/project/def-careg421/ruizhang/2021_guiyang/hmm/${gene}.hmm

# File paths
annotation=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-annotation
featureCounts=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/individual-assembly-featureCounts
output=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output

# Similar procedures as other genes would be applied here
# BLAST analysis would be performed following the same procedure as other genes
```

## 4. Mmseqs

### 4.1 Setup database

```bash
# Setup MMseqs2 database for taxonomy analysis using GTDB
module load StdEnv/2023
module load mmseqs2/15-6f452
module load blast+/2.14.1

cd /project/def-warren21/ruizhang/software/mmseq2_db
INPUT=/project/def-warren21/ruizhang/software/mmseq2_db
mmseqs taxonomy $INPUT/test/mcrA_test_queryDB $INPUT/GTDB $INPUT/test/taxonomyResult1 tmp --lca-ranks species,family
mmseqs createtsv $INPUT/test/mcrA_test_queryDB $INPUT/test/taxonomyResult1 $INPUT/test/mcrA_test.taxonomyResult1.tsv
```

### 4.2 mcrA

```bash
module load StdEnv/2023
module load mmseqs2/15-6f452
module load blast+/2.14.1

# Step1: creating a query DB for the input sequence
gene='mcrA_Pfam'
input=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output
seqTaxDB=/project/def-warren21/ruizhang/software/mmseq2_db
queryDB=$SCRATCH/mmseq2_queryDB/

mmseqs createdb \
$input/AA_gy21_${gene}_combined_sequence.faa \
$queryDB/${gene}_queryDB

# Step2: search against the TaxDB
mmseqs taxonomy \
$queryDB/${gene}_queryDB \
$seqTaxDB/GTDB \
$queryDB/${gene}_taxonomyResult \
$SLURM_TMPDIR --lca-mode 4 --lca-ranks species,family \
--threads 32 --split-memory-limit 90G

# Step3: convert to tsv format
output=/project/def-careg421/ruizhang/2021_guiyang/mmseqs2
mmseqs createtsv \
$queryDB/${gene}_queryDB \
$queryDB/${gene}_taxonomyResult \
$output/${gene}_taxonomyResult.tsv
```

### 4.3 pmoA/mmoX

```bash
module load StdEnv/2023
module load mmseqs2/15-6f452
module load blast+/2.14.1

# Step1: creating a query DB for the input sequence
gene='pmoA_graftM'
input=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output
seqTaxDB=/project/def-warren21/ruizhang/software/mmseq2_db
queryDB=$SCRATCH/mmseq2_queryDB/

mmseqs createdb \
$input/AA_gy21_${gene}_combined_sequence.faa \
$queryDB/${gene}_queryDB

# Step2: search against the TaxDB
mmseqs taxonomy \
$queryDB/${gene}_queryDB \
$seqTaxDB/GTDB \
$queryDB/${gene}_taxonomyResult \
$SLURM_TMPDIR --lca-mode 4 --lca-ranks family \
--threads 32 --split-memory-limit 90G

# Step3: convert to tsv format
output=/project/def-careg421/ruizhang/2021_guiyang/mmseqs2
mmseqs createtsv \
$queryDB/${gene}_queryDB \
$queryDB/${gene}_taxonomyResult \
$output/${gene}_taxonomyResult.tsv

# Similar process can be applied for mmoX gene by changing the gene variable
```

### 4.4 merB

```bash
module load StdEnv/2023
module load mmseqs2/15-6f452
module load blast+/2.14.1

# Step1: creating a query DB for the input sequence
gene='merB'
input=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output
seqTaxDB=/project/def-warren21/ruizhang/software/mmseq2_db
queryDB=$SCRATCH/mmseq2_queryDB/

mmseqs createdb \
$input/AA_gy21_${gene}_combined_sequence.faa \
$queryDB/${gene}_queryDB

# Step2: search against the TaxDB
mmseqs taxonomy \
$queryDB/${gene}_queryDB \
$seqTaxDB/GTDB \
$queryDB/${gene}_taxonomyResult \
$SLURM_TMPDIR --lca-mode 4 --lca-ranks family \
--threads 32 --split-memory-limit 90G

# Step3: convert to tsv format
output=/project/def-careg421/ruizhang/2021_guiyang/mmseqs2
mmseqs createtsv \
$queryDB/${gene}_queryDB \
$queryDB/${gene}_taxonomyResult \
$output/${gene}_taxonomyResult.tsv
```

### 5.5 hgcA

```bash
module load StdEnv/2023
module load mmseqs2/15-6f452
module load blast+/2.14.1

# Step1: creating a query DB for the input sequence
gene='hgcA'
input=/project/def-careg421/ruizhang/2021_guiyang/functional_annotation/output
seqTaxDB=/project/def-warren21/ruizhang/software/mmseq2_db
queryDB=$SCRATCH/mmseq2_queryDB/

# Similar processes as other genes would be applied here
# Complete taxonomy analysis would follow the same procedure as other genes
```
